cogs/embeds.py

```diff
@@ -1,24 +1,142 @@
 import discord
 from discord.ext import commands
+import json
+import re
 
 
 class Embeds(commands.Cog):
+
     def __init__(self, bot: commands.Bot) -> None:
         self.bot = bot
-    
+
     @commands.hybrid_command()
     async def embed(self, ctx: commands.Context):
         """
         Interactively construct a Discord embed.
         """
-        raise NotImplementedError('Command requires implementation and permission set-up.')
-    
+
+        def check(m: discord.Message):
+            return m.channel == ctx.channel and m.author == ctx.author
+
+        def is_skip(msg: discord.Message):
+            return msg.content.strip().lower() == "skip"
+
+        def is_done(msg: discord.Message):
+            return msg.content.strip().lower() == "done"
+
+
+        async def handle_color(embed: discord.Embed):
+            while True:
+                try:
+                    color_msg: discord.Message = await self.bot.wait_for("message", check=check, timeout=timeout)
+                    if is_skip(color_msg):
+                        break
+                    color_str = color_msg.content
+                    embed.color = discord.Color.from_str(color_str)
+                    break
+                except:
+                    await ctx.send("Invalid Color. Please provide the color in hex format (e.g. `#123456`). Use `skip` to skip this.")
+
+        async def handle_img(embed: discord.Embed):
+            img_url_msg: discord.Message = await self.bot.wait_for("message", check=check, timeout=timeout)
+            if is_skip(img_url_msg): return
+            img_url = img_url_msg.content.strip()
+            if img_url.startswith("<"): img_url = img_url[1:]
+            if img_url.endswith(">"): img_url = img_url[:-1]
+            try:
+                embed.set_image(url=img_url)
+            except:
+                print("Error setting image")
+
+        async def handle_fields(embed: discord.Embed):
+            while True:
+                await ctx.send("You are now adding another field. Please use the following format: `\"Title\" \"Value\"`. You can type `done` to complete the embed.")
+                field_msg = await self.bot.wait_for("message", check=check, timeout=timeout)
+                if is_done(field_msg):
+                    break
+                [title, value] = await handle_field(field_msg)
+                title = title[:256]
+                value = value[:1024]
+                embed.add_field(name=title, value=value)
+
+        async def handle_field(field_msg: discord.Message):
+            output = []
+            current = None
+            field_tokens = re.split(" +", field_msg.content.strip())
```

Comment:
Is this equivalent to `field_msg.content.strip().split()`? Using the string method might be clearer. Check the [python documentation](https://docs.python.org/3/library/stdtypes.html#str.split).

In any case, I am worried about this implementation eating up whitespace in the title and value - I would recommend matching to a pattern like `\"(.*)\" \"(.*)\"` using `re.DOTALL`.

Reply:
As far as I could tell, I can't use regex as a separator when using `string.split()`. Hence the use of `re.split()`

As for the implementation, that is how it is currently done in IB.ai. What I am (or well, tried) doing here is:
- Split the entire string based on spaces.
- Looped over the tokens and checked
- if the token starts with " then we append it to the `current` string
- if the token ends with " then we take the current string, remove the quotation marks, and add it to the output array, and set the `current` string to `null`
- otherwise, if the `current` string is not `null`, we append a space and the token.

I tried keeping it similar to the Java version. I'll try to update it with the regex you provided.

Reply:
We should be sticking to IB.ai's implementation of these commands but in this case something like
```py
match = re.search('\"(.*)\" \"(.*)\"', field_msg.content)
if match is None:
    ...  # handle invalid format
title = match.group(1)[:256]
value = match.group(2)[:1024]
```
would be clearer and more pythonic in my opinion. (Although this is not equivalent to the IB.ai `embed` command behavior I think this is natural considering the prompt to send data in the form `"Title" "Value"`, but I would again wait for Ray or Nathaneal to take a closer look at this.)

Reply:
I think I agree, though I'd use a this regex pattern `\"(.+?)\" \"(.+?)\"` and I'd also strip the content, just to be safe.

Reply:
Good catch - if it is a problem (which it is), I reckon we should try fixing it. As much as we should try keeping implementations relatively similar to the original code, we are working with a different language and so clearer code is definitely a plus. 

I'm for using the quoted code with a slight change to the regex to make sure we catch cases where there are multiple whitespaces between the 2 quotes.

> ```python
> match = re.search('\"(.*)\" \"(.*)\"', field_msg.content)
> if match is None:
>     ...  # handle invalid format
> title = match.group(1)[:256]
> value = match.group(2)[:1024]
> ```

Change regex search string to `\"(.*)\" +\"(.*)\"`

Other than that, it looks good to me.